File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.py
from collections import defaultdict
from collections import deque

def solution(begin, target, words):
    answer = -1
    if target not in words:
        return 0
    
    graph = defaultdict(list)
    
    words.append(begin)
    l = len(words[0])
    for word1 in words:
        for word2 in words:
            diff = 0
            for k in range(l):
                if word1[k] != word2[k]:
                    diff += 1
            if diff == 1:
                graph[word1].append(word2)

    visited = {word : False for word in words}
    q = deque([(begin, 0)])
    visited[begin] = True
    while q:
        v, step = q.popleft()
        if v == target:
            return step
        for word in graph[v]:
            if not visited[word]:
                q.append((word, step+1))
                visited[word] = True
    return 0

File: 프로그래머스/3/43163. 단어 변환/단어 변환.js
const diff = (str1, str2) => {
    let d = 0;
    for (let i=0; i<str1.length; i++) {
        if (str1[i] !== str2[i]) d++;
    }
    return d;
}

function solution(begin, target, words) {
    let q = [[begin, 0]]
    let visited = new Array(words.length).fill(false)
    
    while (q.length !== 0) {
        let [s, d] = q.shift()
        if (s === target) return d
        for (const i in words) {
            if (!visited[i] && diff(words[i], s) === 1) {
                visited[i] = true // 방문처리
                q.push([words[i], d+1])
            }
        }
        
    }
    
    return 0;
}